itest/lnd_onchain_test.go

```diff
@@ -731,3 +731,115 @@ func genAnchorSweep(ht *lntest.HarnessTest,
 
 	return btcutil.NewTx(tx)
 }
+
+// testRemoveTx tests that we are able to remove an unconfirmed transaction
+// from the internal wallet as long as the tx is still unconfirmed. This test
+// also verifies that after the tx is removed (while unconfirmed) it will show
+// up as confirmed as soon as the original transaction is mined.
+func testRemoveTx(ht *lntest.HarnessTest) {
+	// Create a new node so that we start with no funds on the internal
+	// wallet.
+	alice := ht.NewNode("Alice", nil)
+
+	const initialWalletAmt = btcutil.SatoshiPerBitcoin
+
+	// Funding the node with an initial balance.
+	ht.FundCoins(initialWalletAmt, alice)
+
+	// Create an address for Alice to send the coins to.
+	req := &lnrpc.NewAddressRequest{
+		Type: lnrpc.AddressType_WITNESS_PUBKEY_HASH,
+	}
+	resp := alice.RPC.NewAddress(req)
+
+	// We send half the amount to that address generating two unconfirmed
+	// outpoints in our internal wallet.
+	sendReq := &lnrpc.SendCoinsRequest{
+		Addr:   resp.Address,
+		Amount: initialWalletAmt / 2,
+	}
+	alice.RPC.SendCoins(sendReq)
+	txID := ht.Miner.AssertNumTxsInMempool(1)[0]
+
+	// Make sure the unspent number of utxos is 2 and the unconfirmed
+	// balances add up.
+	unconfirmed := ht.GetUTXOsUnconfirmed(
+		alice, lnwallet.DefaultAccountName,
+	)
+	require.Lenf(ht, unconfirmed, 2,
+		"expected 2 unconfirmed tx, got %v", len(unconfirmed))
+
+	var unconfirmedAmt btcutil.Amount
+	for _, utxo := range unconfirmed {
+		unconfirmedAmt += btcutil.Amount(utxo.AmountSat)
+	}
+
+	aliceBalResp := alice.RPC.WalletBalance()
+	require.Equal(ht, unconfirmedAmt,
+		btcutil.Amount(aliceBalResp.UnconfirmedBalance))
+
+	// Get the raw transaction to remove it later.
+	tx := ht.Miner.GetNumTxsFromMempool(1)[0]
+	// Calculate the tx fee so we can later compare the end amounts. We
+	// are sending from the internal wallet to the internal wallet so only
+	// the tx fee apply when calucalting the final amount of the wallet.
+	txFee := ht.CalculateTxFee(tx)
+
+	// Now remove the transaction. We should see that the wallet state
+	// equals the amount prior to sending the transaction. It is important
+	// to understand that we do not remove any transaction from the mempool
+	// (thats not possible in reality) we just remove it from our local
+	// store.
+	var buf bytes.Buffer
+	require.NoError(ht, tx.Serialize(&buf))
+	alice.RPC.RemoveTransaction(&walletrpc.Transaction{
+		TxHex: buf.Bytes(),
+	})
+
+	// Verify that the balance equals the initial state.
+	confirmed := ht.GetUTXOsConfirmed(
+		alice, lnwallet.DefaultAccountName,
+	)
+	require.Lenf(ht, confirmed, 1,
+		"expected 1 confirmed tx, got %v", len(confirmed))
+
+	var confirmedAmt btcutil.Amount
+	for _, utxo := range confirmed {
+		confirmedAmt += btcutil.Amount(utxo.AmountSat)
+	}
+
+	aliceBalResp = alice.RPC.WalletBalance()
+	require.Equal(ht, confirmedAmt,
+		btcutil.Amount(aliceBalResp.ConfirmedBalance))
```

Comment:
so the confirmed balance is zero here?

Reply:
reworked the tests because of your good comment, so that I am now comparing real amounts, so in this case it will be `Initial_Amount - TxFee`